src/rules/noShadowedVariableRule.ts

```diff
@@ -38,117 +39,137 @@ export class Rule extends Lint.Rules.AbstractRule {
         }
-
-        super.visitVariableDeclaration(node);
     }
 
-    private handleSingleVariableIdentifier(variableIdentifier: ts.Identifier, isBlockScoped: boolean) {
-        const variableName = variableIdentifier.text;
-
-        if (this.isVarInCurrentScope(variableName) && !this.inCurrentBlockScope(variableName)) {
-            // shadowing if there's already a `var` of the same name in the scope AND
-            // it's not in the current block (handled by the 'no-duplicate-variable' rule)
-            this.addFailureOnIdentifier(variableIdentifier);
-        } else if (this.inPreviousBlockScope(variableName)) {
-            // shadowing if there is a `var`, `let`, 'const`, or parameter in a previous block scope
-            this.addFailureOnIdentifier(variableIdentifier);
-        }
-
-        if (!isBlockScoped) {
-            // `var` variables go on the scope
-            this.getCurrentScope().add(variableName);
+    private handleBindingName(node: ts.BindingName, blockScoped: boolean) {
+        if (node.kind === ts.SyntaxKind.Identifier) {
+            this.scope.addVariable(node, blockScoped);
+        } else {
+            for (const element of node.elements) {
+                if (element.kind !== ts.SyntaxKind.OmittedExpression) {
+                    this.handleBindingName(element.name, blockScoped);
+                }
+            }
         }
-        // all variables go on block scope, including `var`
-        this.getCurrentBlockScope().add(variableName);
-    }
-
-    private isVarInCurrentScope(varName: string) {
-        return this.getCurrentScope().has(varName);
     }
 
-    private inCurrentBlockScope(varName: string) {
-        return this.getCurrentBlockScope().has(varName);
-    }
-
-    private inPreviousBlockScope(varName: string) {
-        return this.getAllBlockScopes().some((scopeInfo) => {
-            return scopeInfo !== this.getCurrentBlockScope() && scopeInfo.has(varName);
+    private onScopeEnd(parent?: Scope) {
+        const {variables, variablesSeen} = this.scope;
+        variablesSeen.forEach((identifiers, name) => {
+            if (variables.has(name)) {
+                for (const identifier of identifiers) {
+                    this.addFailureAtNode(identifier, Rule.FAILURE_STRING_FACTORY(name));
+                }
+            } else if (parent !== undefined) {
+                addToList(parent.variablesSeen, name, identifiers);
+            }
         });
+        if (parent !== undefined) {
+            variables.forEach((identifiers, name) => {
+                addToList(parent.variablesSeen, name, identifiers);
+            });
+        }
     }
+}
 
-    private addFailureOnIdentifier(ident: ts.Identifier) {
-        const failureString = Rule.FAILURE_STRING_FACTORY(ident.text);
-        this.addFailureAtNode(ident, failureString);
+function addToList(map: Map<string, ts.Identifier[]>, name: string, identifiers: ts.Identifier[]) {
```

Comment:
This function might need to be used elsewhere ([multimap](http://www.cplusplus.com/reference/map/multimap/)s show up often), but with a different type of values.